CORE-14251 Client Monitoring: Implement a cluster-wide endpoint for client connections

pgellert opened the pull request:
This endpoint is to provide a cluster-wide aggregated view of the
per-broker ListKafkaConnection endpoint. This is to simplify rpk's logic
for surfacing a cluster-wide view of "top" kafka connections.

See the commits for implementation details.

Fixes https://redpandadata.atlassian.net/browse/CORE-14251

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
proto: minor improvements to ListKafkaConnections

- fix request_count_total field name in doc comment
- add doc comment to request-response messages
- extract the kafka_connections proto into a separate library for easier
  sharing

pgellert committed:
admin: extract kafka_connections_service

To encapsulate the logic of konnection gathering.

The code is only moved, the only changes are extracting a
`get_effective_limit` function for later reuse, and some minor fixups to
make the code compile (e.g.; type name adjustments due to moving across
namespaces).

pgellert committed:
admin: extract license utils

Extract it to a utils file to allow sharing it with other admin
endpoints as well later.

pgellert committed:
admin: refactor connection_collector interface

Rename extract --> extract_unordered()
Rename ordered_collector::extract_sorted --> extract()
Define unordered_collector::extract --> same as extract_unordered()

extract() now means extract as per the oredered'ness of the collector,
while extract_unordered() may be unordered even for the
ordered_collector.

This enabled more code reuse across the ordered and unordered paths, and
will also allow more code reuse in the cluster-level endpoint.

pgellert committed:
admin: use optimized add_all method

This is to allow using the faster `async_push_range` function.

rockwotj commented on proto/redpanda/core/admin/v2/broker.proto:
```diff
@@ -14,13 +14,13 @@
 syntax = "proto3";
 
-package redpanda.core.admin.v2;
+package redpanda.core.admin.v2.broker_service;
```
This should match the directory and version number is supposed to go last.

pgellert replied:
I opted to back out of the namespacing and use an optional node_id field on the request instead, as discussed.

rockwotj commented on proto/redpanda/core/admin/v2/cluster.proto:
```diff
@@ -0,0 +1,101 @@
+
+syntax = "proto3";
+
+package redpanda.core.admin.v2.cluster_service;
```
Same note as before

michael-redpanda commented on proto/redpanda/core/admin/v2/cluster.proto:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+message ListKafkaConnectionsRequest {
```
@rockwotj should all messages also have an overall comment?  I realize this one is probably self explanatory, but just asking

rockwotj replied:
Yeah, we want all protobuf to have documentation even if it seems self-explanatory (because it's probably not for users).

Please follow the guide Kat put together: https://redpandadata.slack.com/archives/C09C27KTCPK/p1759766098707339

pgellert replied:
I've added a comment here now as part of the first commit.

pgellert committed:
admin: ListKafkaConnections as ClusterService rpc

Move the endpoint from the BrokerService to the ClusterService, as we
intend to make this endpoint return cluster-wide data to simplify rpk's
logic of surfacing a cluster-wide view of "top" kafka connections.

As part of this, the node_id field on the request is removed. The
endpoint is always going to return cluster-wide data, and internal
gathering is going to be implemented using `ctx.is_proxied()` in the
next comimt.

The is purely a code-shifting commit. The collection of cluster-wide
data is implemented in the next commit.

pgellert committed:
admin: implement cluster-wide collection of ListKafkaConnections

This endpoint queries all of the brokers one-by-one and uses the defined
ordering to combine the results.

The ducktape test is made multi-node now that connections are gathered
across the whole cluster. The timeout is bumped to 30s because I saw
some flakiness at the 15s timeout (~1/10 failures).

pgellert commented:
force-push:
* [first](https://github.com/redpanda-data/redpanda/compare/ffef9fee63be48a76cc7f6351f9097b7948320d9..db607a04f065d82ba4dac3ed81b0503292bf2014): address code review feedback
  - back out of the complexity of namespacing the BrokerService + ClusterService, and instead, have a single ClusterService endpoint with an `optional int32 node_id` parameter
  - add missing doc comment
  - fix test name typo
* [second](https://github.com/redpanda-data/redpanda/compare/db607a04f065d82ba4dac3ed81b0503292bf2014..3fc80958fe247706e4c6db9456ea71b24f02e5be): rebase to dev to fix merge conflict

pgellert commented on tests/rptest/tests/list_kafka_connections_test.py:
```diff
@@ -160,6 +160,58 @@ def valid_response() -> bool:
         assert conn.close_time.ToDatetime() > datetime(year=2025, month=1, day=1)
 
+    @cluster(num_nodes=4)
+    def test_cluser_service_list_kafka_connections(self):
```
Fixed

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/3fc80958fe247706e4c6db9456ea71b24f02e5be..e16d32ec00ee36fcffb1a059f332371838c5e514): address copilot code review

pgellert commented on src/v/redpanda/admin/services/utils.h:
```diff
@@ -0,0 +1,18 @@
+ * by the Apache License, Version 2.0
+ */
+
+#include "features/feature_table.h"
```
Fixed

pgellert commented on proto/redpanda/core/admin/v2/cluster.proto:
```diff
@@ -0,0 +1,110 @@
+    string order_by = 4;
+}
+
+// ListKafkaConnectionsRequest is the response from the ListKafkaConnections
```
Fixed

BenPope commented on src/v/redpanda/admin/services/broker.cc:
```diff
@@ -95,128 +88,6 @@ proto::admin::broker broker_service_impl::self_broker() const {
 
 namespace {
 
-struct connection_collector {
```
The commit message says "copy and pasted", but this is all just moved, right?

pgellert replied:
Right, yeah, let me reword to "moved".

BenPope commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -110,39 +113,44 @@ kafka_connections_service::list_kafka_connections_local(
+
+            auto make_local_collector = [limit](size_t accumulated_count) {
+                return ss::make_shared<unordered_collector>(
+                  limit - accumulated_count);
```
It's not obvious to me why these are `shared_ptr`; is it to avoid taking/capturing references in a coroutine?

pgellert replied:
As discussed, it's to be able to pass around a pointer to a polymorphic `connection_collector`. I've replaced the `ss::make_shared` with a `std::unique_ptr` though as it doesn't need to be shared.

BenPope commented on src/v/redpanda/admin/kafka_connections_service.h:
```diff
@@ -47,6 +50,14 @@ class unordered_collector : public connection_collector {
+        _connections.reserve(_connections.size() + conns.size());
+        co_await ssx::async_for_each(std::move(conns), [this](auto& conn) {
+            _connections.emplace_back(std::move(conn));
+        });
```
This does not respect `_limit` like `add` and `bounded_priority_queue` do, is that intentional? I think you need to subrange `conns`.

pgellert replied:
No, good catch, thanks. I've fixed this now.

BenPope commented on proto/redpanda/core/admin/v2/cluster.proto:
```diff
@@ -0,0 +1,110 @@
+    // If specified, the results are returned only for the broker with the
+    // specified node ID. If set to -1, the broker handling the RPC request
+    // returns information about itself.
+    optional int32 node_id = 1;
```
I prefer this optional as unification of broker/cluster as to what was here previously.

Since I see scope for this being a common pattern (map brokers, reduce) pattern, should there be something even more explicit and used by other endpoints? @rockwotj

pgellert replied:
> should there be something even more explicit

What do you have in mind in terms of something more explicit?

rockwotj replied:
I would be up for making the requests use `optional` instead of `-1`, however here `optional` unset means that we want all brokers. TBH, I would rather remove this, rpk and console will always want a cluster wide view and I think we can optimize the filter to target specific nodes later on.

rockwotj replied:
I don't expect to have many broker service endpoints, we should strongly be pushing for cluster wide endpoints.

BenPope replied:
> > should there be something even more explicit
> 
> What do you have in mind in terms of something more explicit?

I'm not really familiar with best practice here, but to me, sentinel values like -1, even with documentation, are just awkward. Perhaps some kind of "node_list OR cluster".

pgellert replied:
I've moved to removing the `node_id` field now. As discussed with @rockwotj, and as the TODO comment I added states, we can optimize later by parsing the `filter` string and not querying other brokers if we safely determine that  that is unnecessary. We can rely on `ctx.is_proxied()` to implement the scatter-gather logic.

BenPope commented on src/v/redpanda/admin/services/cluster.cc:
```diff
@@ -39,6 +41,74 @@ cluster_service_impl::cluster_service_impl(
+
+    auto local_resp = co_await _kafka_connections_service.local()
+                        .list_kafka_connections_local(make_broker_req(-1));
+    co_await add_to_response(std::move(local_resp));
```
If there is an unordered limit, does it make sense to gather locally first?

pgellert replied:
We would still need to send a request to other nodes to collect the total count, but we could optimize by reducing the limit sent to other nodes. I haven't implemented that yet but we can keep that in mind as a follow up optimization.

BenPope commented on tests/rptest/tests/list_kafka_connections_test.py:
```diff
@@ -160,6 +160,58 @@ def valid_response() -> bool:
         assert conn.close_time.ToDatetime() > datetime(year=2025, month=1, day=1)
 
+    @cluster(num_nodes=4)
+    def test_cluster_service_list_kafka_connections(self):
```
Is it worth factoring out the overlap with `test_list_kafka_connections`?

pgellert replied:
With the `node_id` field removed, the tests became simpler here so this no longer applies

rockwotj commented on src/v/redpanda/admin/kafka_connections_service.h:
```diff
@@ -0,0 +1,98 @@
+
+namespace admin {
+
+struct connection_collector {
```
Can't these go into the implementation file instead of the header?

pgellert replied:
Yes, I've reworked this now

rockwotj commented on src/v/redpanda/admin/kafka_connections_service.cc:
> extract_unordered() may be unordered even for the unordered_collector

You mean `ordered_collector`?

pgellert replied:
Yes, updated

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f8c417bb-554a-439d-9ab8-98178539d6c1-->
<!--TEST RESULTS-->
<!--19e4931f-1402-4ab4-81e0-4b58fb07f981-->
<!--TEST RESULTS-->
<!--c36611bb-e247-4f9c-b880-de4a6574fff8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74682">build#74682</a></summary>


| test_class                 | test_method                     | test_arguments        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                       |
|:---------------------------|:--------------------------------|:----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests       | test_rapid_shadow_link_toggling | null                  | integration | https://buildkite.com/redpanda/redpanda/builds/74682#019a0be2-f949-4e10-950a-3b9a1052e624 | FLAKY         | 17/21    | upstream reliability is '94.65737514518003'. current run reliability is '80.95238095238095'. drift is 13.70499 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
| PartitionMovementTest      | test_empty                      | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74682#019a0be2-f94c-4916-ac12-a287f1d9365e | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                     |
| ShadowLinkingRandomOpsTest | test_node_operations            | {"failures": false}   | integration | https://buildkite.com/redpanda/redpanda/builds/74682#019a0be2-f947-4214-9882-df5f2c8a9ec3 | FLAKY         | 10/21    | upstream reliability is '35.66433566433567'. current run reliability is '47.61904761904761'. drift is -11.95471 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations      |
| ShadowLinkingRandomOpsTest | test_node_operations            | {"failures": true}    | integration | https://buildkite.com/redpanda/redpanda/builds/74682#019a0be2-f948-47ed-821c-38a6a8771802 | FLAKY         | 4/21     | upstream reliability is '54.385964912280706'. current run reliability is '19.047619047619047'. drift is 35.33835 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations      |
| ShadowLinkingRandomOpsTest | test_node_operations            | {"failures": true}    | integration | https://buildkite.com/redpanda/redpanda/builds/74682#019a0be6-3fd6-45ee-b398-65bcc5b8e383 | FLAKY         | 6/21     | upstream reliability is '54.385964912280706'. current run reliability is '28.57142857142857'. drift is 25.81454 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74704">build#74704</a></summary>


| test_class                      | test_method              | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMovementTest           | test_static              | {"num_to_upgrade": 0}                              | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1e-109d-4de9-b1a9-478bbf840d3f | FLAKY         | 18/21    | upstream reliability is '99.06542056074767'. current run reliability is '85.71428571428571'. drift is 13.35113 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                        |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1d-014d-456b-9684-24758277f638 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| ShadowLinkingRandomOpsTest      | test_node_operations     | {"failures": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1d-014d-456b-9684-24758277f638 | FLAKY         | 9/21     | upstream reliability is '37.301587301587304'. current run reliability is '42.857142857142854'. drift is -5.55556 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations          |
| ShadowLinkingRandomOpsTest      | test_node_operations     | {"failures": false}                                | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1e-109f-4b2b-90c6-591c49ca6ee2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations          |
| ShadowLinkingRandomOpsTest      | test_node_operations     | {"failures": true}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1d-0145-4f81-94c0-2df6f3ce2417 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations          |
| ShadowLinkingRandomOpsTest      | test_node_operations     | {"failures": true}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74704#019a0d1e-1096-49da-abc1-2db8802d5ac4 | FLAKY         | 1/21     | upstream reliability is '33.5548172757475'. current run reliability is '4.761904761904762'. drift is 28.79291 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74778">build#74778</a></summary>


| test_class                 | test_method          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74778#019a10b2-02d2-414c-91e9-7d1d847b822d | FLAKY         | 3/21     | upstream reliability is '36.23804463336875'. current run reliability is '14.285714285714285'. drift is 21.95233 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74778#019a10b2-02cd-40bc-a635-d149b144b6b9 | FAIL          | 0/1      |                                                                                                                                                                          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74778#019a10b2-02d2-414c-91e9-7d1d847b822d | FLAKY         | 3/21     | upstream reliability is '31.11111111111111'. current run reliability is '14.285714285714285'. drift is 16.8254 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<!--/body-->

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/e16d32ec00ee36fcffb1a059f332371838c5e514..67e0b5dd03cba89036c23df3b9b62fb31af11036): address review feedback
* Remove `node_id` from the cluster-wide endpoint, and rely on the `ctx.is_proxied()` field for implementing the scatter-gather logic. Rework the ducktape tests a bit as part of this.
* Reduce the amount of code in the `kafka_connection_service.h` header, and move it into `kafka_connection_service.cc`
* Fix the `unordered_collector::add_all` bug
* `ss::shared_ptr` -> `std::unique_ptr` for `connection_collector` types

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/67e0b5dd03cba89036c23df3b9b62fb31af11036..54581e69ecd060ebe49b1446ad02d380c3980be4): some more minor improvements based on feedback + self-review

rockwotj commented on tests/rptest/tests/list_kafka_connections_test.py:
```diff
@@ -123,7 +123,7 @@ def valid_response() -> bool:
         wait_until(
             valid_response,
-            timeout_sec=15,
+            timeout_sec=30,
```
little concerning we couldn't do all this within 15 seconds...

BenPope commented on src/v/redpanda/admin/kafka_connections_service.cc:
```diff
@@ -53,6 +56,18 @@ class unordered_collector : public connection_collector {
+    add_all(chunked_vector<proto::admin::kafka_connection> conns) final {
+        auto to_add_count = std::min(
+          conns.size(), _limit - _connections.size());
+        auto insert_range = std::ranges::subrange(
+          conns.begin(), conns.begin() + to_add_count);
```
nitpick: As of https://github.com/redpanda-data/redpanda/pull/28125 you can do:
```suggestion
        auto insert_range = conns | std::views::take(to_add_count);
```

BenPope reviewed (dismissed):
Very nice. No need to address the nitpick, it was more a PSA.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74704-->
<!--fbe6f4d9-cedd-40e8-ac59-6a565f28cc8e-->
<!--967523d6-b786-4716-a1fd-e64f405214f7-->
<!--4620a83d-92ae-4ea5-b72b-4bcf3fabccc8-->
<!--title-->
## Retry command for [Build#74704](https://buildkite.com/redpanda/redpanda/builds/74704)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

BenPope commented:
/ci-repeat 1
tests/rptest/tests/random_node_operations_smoke_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/54581e69ecd060ebe49b1446ad02d380c3980be4..743657b7bfc45542673b22700db4aa7e06137181): rebase to dev to fix merge conflict on `tests/rptest/clients/admin/proto/redpanda/core/admin/v2/shadow_link_pb2.pyi` which was stale on dev earlier

BenPope commented:
> [force-push](https://github.com/redpanda-data/redpanda/compare/54581e69ecd060ebe49b1446ad02d380c3980be4..743657b7bfc45542673b22700db4aa7e06137181): rebase to dev to fix merge conflict on `tests/rptest/clients/admin/proto/redpanda/core/admin/v2/shadow_link_pb2.pyi` which was stale on dev earlier

This merge conflict seems strange to me, since the commit didn't go near it, as far as I can tell.

pgellert commented:
It did before. That's a generated file, and it was stale on dev because someone must have forgotten to run `task rp:generate-ducktape-protos` and then my old commit included some generated proto changes when I ran the generator. I guess it got updated on dev in the meantime, which is where the conflict was.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74775-->
<!--d63f1648-2b62-4ea3-a238-7cd5acf69152-->
<!--f61903a1-a1a1-4750-8694-322c90d22731-->
<!--3585efae-31b1-4200-823c-b72688eb6ccb-->
<!--title-->
## Retry command for [Build#74775](https://buildkite.com/redpanda/redpanda/builds/74775)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

pgellert commented:
/ci-repeat 1
skip-unit
skip-redpanda-build
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74778-->
<!--7c6fe273-934b-4a03-b5bf-512d230a168d-->
<!--title-->
## Retry command for [Build#74778](https://buildkite.com/redpanda/redpanda/builds/74778)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

pgellert commented:
/test-release-pipeline
